### PR TITLE
Push history when navigating to route-not-found view

### LINKED ISSUE #8544: Browser history not updated for broken router links in 14.2.x
### Description of the bug
Before Vaadin 14.2.0 following a broken router link caused a browser history update followed by the 404/NotFound view.

Since 14.2.0 the 404/NotFound view is still shown but the browser history is not updated. This causes bad UX as 
- the browser URL is not updated
- using browser "back" button results in unexpected page

### Minimal reproducible example
I was able to reproduce the issue by downloading the latest REST starter from https://vaadin.com/start/v14/rest-example and updating one of the router links in the main view via the browser dev tools. Clicking on this broken link causes the behavior described above.

### Versions:
- Vaadin 14.2.x
  - regression caused by fix for https://github.com/vaadin/flow/issues/3619
- all browsers

### PR DESCRIPTION
**Pull request by joheriks**
This fixes a regression introduced with #8125 . 
Fixes #8544

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java [hunk 1]
```diff
@@ -243,11 +244,21 @@ public int handle(NavigationEvent event) {
...
     private void pushHistoryStateIfNeeded(NavigationEvent event, UI ui) {
-        if (event instanceof ErrorNavigationEvent) {
-            return;
-        }
+        boolean isRouterLink = NavigationTrigger.ROUTER_LINK
+                .equals(event.getTrigger());
 
-        if (NavigationTrigger.ROUTER_LINK.equals(event.getTrigger())) {
+        if (event instanceof ErrorNavigationEvent) {
+            if (isRouterLink) {
```
Comment (denis-anisimov) [inline flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java:251]: I would suggest combine `(event instanceof ErrorNavigationEvent) && isRouterLink` in  one `if`.
Less nested ifs.

File: flow-tests/test-root-context/src/test/java/com/vaadin/flow/uitest/ui/BrokenRouterLinkIT.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        WebElement link = findElement(By.id(BrokenRouterLinkView.LINK_ID));
+
+        String href = link.getAttribute("href");
+
+        link.click();
+
+        Assert.assertTrue(getDriver().getCurrentUrl().endsWith(href));
+
+    }
+}
```
Comment (denis-anisimov) [inline flow-tests/test-root-context/src/test/java/com/vaadin/flow/uitest/ui/BrokenRouterLinkIT.java:41]: new line

File: flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java (hunk 1 above)
Comment (joheriks) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java]: Also the other conditions should not be tested at all if event is of type `ErrorNavigationEvent`. Simplified the if structure and extracted the logic to another method.

Comment (knoobie) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java]: You can combine everything inside the method - java uses short-circuit evaluation if used with `&&` or `||`, meaning the right side won't be executed, when left already matched.

File: flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java [hunk 3]
```diff
@@ -284,6 +290,14 @@ protected boolean shouldPushHistoryState(NavigationEvent event) {
         return NavigationTrigger.UI_NAVIGATE.equals(event.getTrigger());
     }
 
+    protected boolean shouldPushHistoryStateForNavigationError(
```
Comment (denis-anisimov) [inline flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java:293]: I think we require javadocs for `protected` methods.

I would avoid `should`  in method names. Not sure whether this is a better name though.

Comment (joheriks) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java]: No need for `protected` as a matter of fact, made `private`. Also renamed.

**Pull request merged**